File: backtracking/0784_letter_case_permutation/0784_letter_case_permutation.py
from typing import List

class Solution:
    def letterCasePermutation(self, S: str) -> List[str]:
        ans = [[]]

        for char in S:
            n = len(ans)
            if char.isalpha():
                for i in range(n):
                    ans.append(ans[i][:])
                    ans[i].append(char.lower())
                    ans[n+i].append(char.upper())
            else:
                for i in range(n):
                    ans[i].append(char)

        return map("".join, ans)


S = "a1b2"
res = Solution().letterCasePermutation(S)
print(res)

File: backtracking/0784_letter_case_permutation/0784_letter_case_permutation.go
package main

import "fmt"

func letterCasePermutation(S string) []string {
	permutations := []string{S}
	for i, v := range S {
		if v >= 'a' && v <= 'z' {
			length := len(permutations)
			for ip := 0; ip < length; ip++ {
				b := []byte(permutations[ip])
				b[i] = byte(v + 'A' - 'a')
				permutations = append(permutations, string(b))
			}
		} else if v >= 'A' && v <= 'Z' {
			length := len(permutations)
			for ip := 0; ip < length; ip++ {
				b := []byte(permutations[ip])
				b[i] = byte(v + 'a' - 'A')
				permutations = append(permutations, string(b))
			}
		}
	}
	return permutations
}

func main() {
	S := "a1b2"
	res := letterCasePermutation(S)
	fmt.Println(res)
}
